File: algo/interface.cpp

#include "algo/interface.h"

c_bitmap * create(unsigned height, unsigned width)
{
	return new c_bitmap(height, width);
}

double get(c_bitmap * bmp, unsigned y, unsigned x, unsigned ch_num)
{
	return bmp->pixel(y, x).get(ch_num);
}

void set(c_bitmap * bmp, unsigned y, unsigned x, unsigned ch_num, double value)
{
	bmp->pixel(y, x).set(ch_num, value);
}

unsigned height(c_bitmap * bmp)
{
	return bmp->height;
}

unsigned width(c_bitmap * bmp)
{
	return bmp->width;
}

unsigned ul_y(c_bitmap * bmp)
{
	return bmp->aoi().ul_y();
}

unsigned ul_x(c_bitmap * bmp)
{
	return bmp->aoi().ul_x();
}

unsigned lr_y(c_bitmap * bmp)
{
	return bmp->aoi().lr_y();
}

unsigned lr_x(c_bitmap * bmp)
{
	return bmp->aoi().lr_x();
}

unsigned aoi_height(c_bitmap * bmp)
{
	return bmp->aoi().height();
}

unsigned aoi_width(c_bitmap * bmp)
{
	return bmp->aoi().width();
}

File: algo/interface/interface.hs

{-# LANGUAGE ForeignFunctionInterface #-}

module Interface where

import Foreign
import Foreign.Ptr()
import Foreign.C.Types

type Bitmap = StablePtr ()

foreign import ccall safe "interface.h"
	create :: CUInt -> CUInt -> Bitmap

foreign import ccall safe "interface.h"
	get :: Bitmap -> CUInt -> CUInt -> CUInt -> CDouble

foreign import ccall safe "interface.h"
	set :: Bitmap -> CUInt -> CUInt -> CUInt -> CDouble -> IO()

foreign import ccall safe "interface.h"
	height :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	width :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	ul_y :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	ul_x :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	lr_y :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	lr_x :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	aoi_height :: Bitmap -> CUInt

foreign import ccall safe "interface.h"
	aoi_width :: Bitmap -> CUInt

